### PR TITLE
opt ce loss

### PR DESCRIPTION
**Pull request by GeeeekExplorer**
As shown in the following code, the current cross entropy loss computation is very slow. The results of the code are 1.2s and 0.02s respectively.

```python
import time
import torch
import torch.nn.functional as F

def bench(func):
    for _ in range(3):
        func()
    torch.cuda.synchronize()
    t = time.time()
    for _ in range(10):
        func()
    torch.cuda.synchronize()
    return time.time() - t

batch_size = 4
seq_len = 2048
vocab_size = 100 * 1024
output_tensor = torch.randn(batch_size, seq_len, vocab_size, dtype=torch.bfloat16, device="cuda")
target_ids = torch.randint(0, vocab_size, (batch_size, seq_len), dtype=torch.long, device="cuda")

print(bench(lambda: F.cross_entropy(output_tensor.transpose(1, 2), target_ids, reduce='mean')))

print(bench(lambda: F.cross_entropy(output_tensor.flatten(0, 1), target_ids.flatten(), reduce='mean')))
```

In this PR, I replace the cross entropy loss computation. In my experiment of `Qwen2.5-0.5B`, MFU (Model FLOPs Utilization) of dp4pp2tp1 increased from 5% to 40%.

### CODE REVIEW

**Comment by GeeeekExplorer** [comment]
In my experiment of Qwen3-8B, MFU (Model FLOPs Utilization) of dp4pp2tp1(mbs=4) increased from 34% to 54%. 

@3outeille @zzhhjjj could you take a look? I think it is not a small optimization.

**Comment by zzhhjjj** [comment]
Thank you for the PR. It's very impressive!
